core/src/main/java/org/apache/iceberg/io/DeleteSchemaUtil.java

```diff
@@ -29,7 +29,7 @@ private static Schema pathPosSchema(Schema rowSchema) {
     return new Schema(
         MetadataColumns.DELETE_FILE_PATH,
         MetadataColumns.DELETE_FILE_POS,
-        Types.NestedField.required(
+        Types.NestedField.optional(
```

Comment:
This was necessary for the writer to allow writing position deletes with a row, but still be ok when there is no row.

Currently, the writer code either uses a schema with required "row" field, or a schema without the "row" field.  The former is actually never used.  So changing to optional should have no impact

Reply:
Update: Looks like a few GenericWriter depend on old schema to throw exception if null rows passed in.  This is a change of behavior , but backward compatible.

Reply:
I have some concerns about this change, the `Delete Formats` spec says the reason why this column type should be `required` is to make sure the statistics of the deleted row values is accurate. I think the reason to make sure the statistics are accurate is because of the manifest reader will use them to filter delete files: 
![image](https://user-images.githubusercontent.com/42907416/223306413-38ea612e-8c76-44af-8ee4-740363d75db7.png)
So I think if this type is changed to optional, the statistics will become unreliable, which may cause the delete manifest entry to be incorrectly filtered? This is just my understanding of the spec, but I'm not sure.

Reply:
Yea you are right, this is tricky.  It says this in spec:

> 

2147483544 row | required struct<...> [1] | Deleted row values. Omit the column when not storing deleted rows.
-- | -- | --

> When present in the delete file, row is required because all delete entries must include the row values.

So either, entire position delete file has 'row', or entire file does not have 'row'.  (Currently it seems Spark does not set 'row' at all, ref:  https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeltaWrite.java#L436)

I somehow need a way, when compacting delete files, to know whether the original position file all have rows or not.  I am not sure at the moment how to get this

Reply:
I updated the pr with a fix for this , the idea from chatting with @RussellSpitzer offline.  I made SparkWrite.DeleteWriter now a fan-out that can redirect deletes to two files, one that either has 'row' as required struct, or no 'row' at all.  In most case only one will be chosen.  Thanks for the initial comment.